librad/src/git/storage.rs

```diff
@@ -136,51 +136,106 @@ impl Storage {
     pub fn has_commit<Oid>(&self, urn: &Urn, oid: Oid) -> Result<bool, Error>
     where
         Oid: AsRef<git2::Oid> + Debug,
     {
-        let oid = oid.as_ref();
-        if oid.is_zero() {
-            // XXX: should this be a panic or error?
-            tracing::warn!("zero oid");
-            return Ok(false);
-        }
+        let oid = match self.has_object(oid)? {
+            Some(oid) => oid,
+            None => return Ok(false),
+        };
 
         self.backend
-            .find_commit(*oid)
+            .find_commit(oid)
             .and_then(|commit| {
-                let namespace = Namespace::from(urn);
-                let branch = {
-                    let path = match &urn.path {
-                        Some(refl) => refl.as_str(),
-                        None => "rad/id",
-                    };
-                    path.strip_prefix("refs/").unwrap_or(path)
+                let tip = match self.get_urn_oid(urn)? {
+                    Some(tip) => tip,
+                    None => return Ok(false),
                 };
-                self.backend
-                    .refname_to_id(&format!("refs/namespaces/{}/refs/{}", namespace, branch))
-                    .and_then(|tip| {
-                        Ok(tip == commit.id()
-                            || self.backend.graph_descendant_of(tip, commit.id())?)
-                    })
-                    .or_matches(is_not_found_err, || Ok(false))
+                Ok(tip == commit.id() || self.backend.graph_descendant_of(tip, commit.id())?)
+            })
+            .or_matches(is_not_found_err, || {
+                tracing::debug!("commit not found");
+                Ok(false)
             })
-            .or_matches(is_not_found_err, || Ok(false))
+    }
+
+    /// Check the existence of `oid` as a **tag**.
+    ///
+    /// The result will be `false` if:
+    ///
+    /// 1. No tag could be found for `oid`
+    /// 2. The reference path for the `urn` could not be found (it defaults to
+    /// `rad/id` if not provided)
+    /// 3. The SHA of the tag was not the same as the resolved reference
+    /// 4. The `oid` was the [`zero`][`git2::Oid::zero`] SHA.
+    #[tracing::instrument(level = "debug", skip(self, urn), fields(urn = %urn), err)]
+    pub fn has_tag<Oid>(&self, urn: &Urn, oid: Oid) -> Result<bool, Error>
+    where
+        Oid: AsRef<git2::Oid> + Debug,
+    {
+        let oid = match self.has_object(oid)? {
+            Some(oid) => oid,
+            None => return Ok(false),
+        };
+
+        self.backend
+            .find_tag(oid)
+            .and_then(|tag| {
+                let tip = match self.get_urn_oid(urn)? {
+                    Some(tip) => tip,
+                    None => return Ok(false),
+                };
+                Ok(tip == tag.id())
+            })
+            .or_matches(is_not_found_err, || {
+                tracing::debug!("tag not found");
+                Ok(false)
+            })
+    }
```

Comment:
You're accessing the backend three times here -- why? I think you want to just find the tip of the ref which corresponds to the URN's path, check that the object is of the right type, and then compare the Oids.

Please also note that the tag name (as recorded in the tag object) is not the same as the ref name.

Reply:
> Please also note that the tag name (as recorded in the tag object) is not the same as the ref name.

I'm not following what you're alluding to here :thinking: The URN should resolve to a tip and we get the tag via that, so where is the tag name involved?

Reply:
Something for @geigerzaehler to consider: normally, tags are in the global namespace of a repository. This means that the tag name (the `tag` header in the object) is equal to the ref (after `tags/`). For example, the tag `v1.0` is usually at `refs/tags/v1.0`. If the tag is GPG-signed, that signature implies that this is the `v1.0` tag, and it shall be at `refs/tags/v1.0`, and there shall be no other `v1.0` tags beside me.

Since we are scoping tags to the peer-id, this is no longer true -- there can be more than one tag object with `tag v1.0` in the repository. `git verify-tag` does _not_ verify that the tag and ref names match, if you care about "tag teleport" attacks, you'd have to incantate `git verify-tag --verbose`, and look at the output. This is obviously not becoming easier to understand when there are N `v1.0` tags in the repo.

**tl;dr** it would be good compare the ref name (acc. to our conventions) and the tag name higher up in the stack, before showing a "merge" button for example.

Reply:
Ach, I see :) @geigerzaehler might be a good idea to add this to any docs you have on this feature so we don't lose track of the info

Reply:
Thanks for the note. I’ll make sure that we check that the `tag` field matches the ref when resolving the merge requests.